packages/graphql-hooks/src/useClientRequest.js

```diff
@@ -20,28 +21,41 @@ function reducer(state, action) {
         ...state,
         loading: true
       }
-    case actionTypes.CACHE_HIT:
-      if (state.cacheHit) {
-        // we can be sure this is the same cacheKey hit
-        // because we dispatch RESET_STATE if it changes
-        return state
-      }
-      return {
-        ...action.result,
-        cacheHit: true,
-        loading: false
-      }
     case actionTypes.REQUEST_RESULT:
       return {
         ...action.result,
-        cacheHit: false,
-        loading: false
+        loading: false,
+        data: !action.result.cacheHit
+          ? prepareData(state.data, action.result.data, action.updateData)
+          : action.result.data
       }
     default:
       return state
   }
 }
 
+function prepareData(currentData, newData, updater) {
+  if (!updater || !currentData) {
+    return newData
+  }
+
+  if (!newData) {
+    return currentData
+  }
+
+  return updater(currentData, newData)
+}
+
+function useDeepCompareCallback(callback, deps) {
+  const ref = React.useRef()
+
+  if (!deepEqual(deps, ref.current)) {
```

Comment:
Isn't this `deepEqual` check going to be very expensive?

Reply:
I don't think so given what we are comparing. And unless we get rid of non-primitive dependencies in `useCallback` calls (`initialOpts`) I think we're stuck to something like this if we want to memoise the `fetchData` callback.

Also, it's possible that the `deepEqual` is cheaper than whatever the user is doing within any hooks that have a `fetchData` that is changing unnecessarily as a dependency.

Reply:
my goal with this PR (changes on top of @olistic suggestions) was to remove `state.data` from useCallback dependencies that can cause expensive comparison. The downside is that caching has to be moved to `GraphQLClient`. I think without `state` in dependency array the array itself is pretty lean: `variables`, `options: { useCache, skipCache, etc. }`.

Reply:
You can also use `fast-deep-equal` instead of `dequal` — about a [33% increase](https://www.npmjs.com/package/dequal#benchmarks) in performance, although at this level I’m not sure it makes a difference.